### PR TITLE
fix: graphback/core - objectId scalar hex string check

### PR DESCRIPTION
**Pull request by Arenukvern**
## Types of changes

What types of changes does your code introduce to Graphback?

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Other (please specify)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aerogear/graphback/blob/master/CONTRIBUTING.md) document.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Verification

Two tests to check in scalars.test.ts;
- should validate objectId as hex string
- should not validate objectId as hex string

## Further comments

Good day.
I would like to add additional objectId scalar check for hex string representation.

**why**: in some cases we can send ObjectId as hex string representation.
In such cases Graphback will fall as it will trying to check object but not string

This fix will add additional check for it.

Thank you for review!

### CODE REVIEW

**Comment by machi1990** [comment]
@Arenukvern thanks for the PR. What issue is this fixing? I noticed some failing tests, I re-launched the build to see if it were some flasky tests - do you mind having a look at them, the failure looked related to the change.

**Comment by Arenukvern** [comment]
> @Arenukvern thanks for the PR. What issue is this fixing? I noticed some failing tests, I re-launched the build to see if it were some flasky tests - do you mind having a look at them, the failure looked related to the change.

Yes, I've noticed it unfortunately after PR and already on it.

Maybe I wrong, but the problem:

If in request/response graphback receiving ObjectId in hex string type then it falls, because  Grapback checks if value is instance of ObjectId but not string type. 
This fix should let Graphback validate hex strings as ObjectId

File: packages/graphback-core/tests/scalars.test.ts [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
+import { isObjectID } from '../src/scalars/objectId';
+
+describe('Graphback scalars', () => {
+  test('should validate objectId as hex string', () => {
+    const objectId = new ObjectId();
+    const hex = objectId.toHexString();
+    const result = isObjectID(hex);
+    expect(result).toBeTruthy();
+  });
+  test('should not validate objectId as hex string', () => {
```
Comment (machi1990) [inline packages/graphback-core/tests/scalars.test.ts:29]:
```suggestion
  test('should return false for non valid objectId hex string', () => {
```

File: packages/graphback-core/src/scalars/objectId.ts [hunk 2]
```diff
@@ -1,13 +1,28 @@
...
+  let isBsonExtObjectID: (val: any) => boolean = () => false;
+
   try {
     const BsonExtObjectID = require('bson-ext').ObjectID;
-    return value instanceof BsonExtObjectID;
+    isBsonExtObjectID = (val: any) => val instanceof BsonExtObjectID;
+    if (isBsonExtObjectID(value)) return true;
+  } catch {}
+
+  // value can be number or string, try to parse
```
Comment (machi1990) [inline packages/graphback-core/src/scalars/objectId.ts:16]: This comment is outdated, maybe remove it all together or update it.

**Review by machi1990 — CHANGES_REQUESTED** [review]
Added a few comments. Otherwise it is good to be merged. 

@Arenukvern based on https://github.com/aerogear/graphback/pull/2276#issuecomment-793879044 do you think you can add an integration test for it? Thanks for the outstanding work.

**Comment by Arenukvern** [comment]
> Added a few comments. Otherwise it is good to be merged.
> 
> @Arenukvern based on [#2276 (comment)](https://github.com/aerogear/graphback/pull/2276#issuecomment-793879044) do you think you can add an integration test for it? Thanks for the outstanding work.

@machi1990  thank you for your review.
Sent commit with all requested changes.

About integration test - do you mean to add test to test described behaviour in file:
`integration/runtime-workflow-mongo.ts`
is it correct?

If so, will do it.

**One question about linter:** 
I've installed graphback via yarn, so everything is building and testing and seems fine.
But when trying to use lint via "llint" in scripts it gives an error:

```bash
eslint -t stylish --project "tsconfig.json"
Invalid option '-t' - perhaps you meant '-c'?
```
Is it working or it is broken?

**Comment by machi1990** [comment]
> > Added a few comments. Otherwise it is good to be merged.
> > @Arenukvern based on [#2276 (comment)](https://github.com/aerogear/graphback/pull/2276#issuecomment-793879044) do you think you can add an integration test for it? Thanks for the outstanding work.
> 
> @machi1990 thank you for your review.
> Sent commit with all requested changes.
> 
> About integration test - do you mean to add test to test described behaviour in file:
> `integration/runtime-workflow-mongo.ts`
> is it correct?

Yes, that will be fine. Thanks.

> If so, will do it.
> 
> **One question about linter:**
> I've installed graphback via yarn, so everything is building and testing and seems fine.
> But when trying to use lint via "llint" in scripts it gives an error:
> 
> ```shell
> eslint -t stylish --project "tsconfig.json"
> Invalid option '-t' - perhaps you meant '-c'?
> ```
> 
> Is it working or it is broken?

`yarn lint` should work fine after everything has been installed.

**Comment by wtrocki** [comment]
@Arenukvern This have been open for long time. We could not merge it because this blocked our integration tests.

**Comment by Arenukvern** [comment]
> @Arenukvern This have been open for long time. We could not merge it because this blocked our integration tests.

Hi @wtrocki yes, sorry for long time.. Unfortunately cannot make integration test for this. 

The initial reason of this PR was to try run GraphBack on frontend with Dexie as database instead of Offix, but unfortunately, because new versions of GraphqlRunner has some specific backend code, I was not able to run it at all:(

**Comment by wtrocki** [comment]
Thank you!

**Pull request closed without merging**